File: src/agent_dump/collect_events.py
"""Visible dialogue extraction and chunking for collect mode."""

from collections.abc import Iterable, Mapping
import re
from typing import Any

from agent_dump.collect_models import CHUNK_TARGET_CHARS, EVENT_EXTRACT_CHAR_BUDGET, CollectEvent
from agent_dump.collect_summary import normalize_text
from agent_dump.message_filter import should_filter_message_for_export
from agent_dump.transcript import read_messages

IGNORABLE_DIALOGUE_PATTERN = re.compile(
    r"(?:hi|hello|thanks|thank you|你好|您好|好的|收到|明白|嗯嗯|ok|okay)[!！,.，。?？\s]*",
    re.IGNORECASE,
)
EVENT_KIND_BY_ROLE = {"user": "user_message", "assistant": "agent_message"}


def extract_collect_events(
    session_data: Mapping[str, Any],
    *,
    char_budget: int = EVENT_EXTRACT_CHAR_BUDGET,
) -> tuple[tuple[CollectEvent, ...], bool]:
    """Extract visible user and agent messages from one normalized session."""
    events: list[CollectEvent] = []
    used_chars = 0
    for transcript_message in read_messages(session_data):
        if should_filter_message_for_export(transcript_message.raw):
            continue

        role = transcript_message.role
        if role not in EVENT_KIND_BY_ROLE:
            continue

        seen_texts: set[str] = set()
        for part_text in transcript_message.visible_texts:
            normalized = normalize_text(part_text)
            identity = normalized.casefold()
            if not normalized or identity in seen_texts or IGNORABLE_DIALOGUE_PATTERN.fullmatch(normalized):
                continue
            seen_texts.add(identity)
            event = CollectEvent(kind=EVENT_KIND_BY_ROLE[role], role=role, text=normalized)
            event_size = len(render_collect_event(event)) + 1
            if used_chars + event_size > char_budget:
                remaining = char_budget - used_chars - (event_size - len(normalized))
                if remaining > 0:
                    events.append(CollectEvent(kind=event.kind, role=role, text=normalized[:remaining]))
                return tuple(events), True
            events.append(event)
            used_chars += event_size

    return tuple(events), False


def render_collect_event(event: CollectEvent) -> str:
    """Serialize one collect event for an LLM data envelope."""
    return f"[{event.kind}] role={event.role} text={event.text}"


def chunk_collect_events(
    events: Iterable[CollectEvent],
    *,
    target_chars: int = CHUNK_TARGET_CHARS,
) -> list[tuple[CollectEvent, ...]]:
    """Chunk events by approximate serialized size."""
    chunks: list[list[CollectEvent]] = []
    current: list[CollectEvent] = []
    current_size = 0

    for event in events:
        event_size = len(render_collect_event(event)) + 1
        if current and current_size + event_size > target_chars:
            chunks.append(current)
            current = []
            current_size = 0
        current.append(event)
        current_size += event_size

    if current:
        chunks.append(current)
    if not chunks:
        return [()]
    return [tuple(chunk) for chunk in chunks]

File: src/collect_events.rs
use crate::collect_model::Event;
use crate::session::SessionData;
use std::collections::HashSet;
use std::sync::LazyLock;

pub fn extract(data: &SessionData) -> (Vec<Vec<Event>>, bool) {
    static IGNORE: LazyLock<regex::Regex> = LazyLock::new(|| {
        regex::Regex::new(r"(?i)^(?:hi|hello|thanks|thank you|你好|您好|好的|收到|明白|嗯嗯|ok|okay)[!！,.，。?？\s]*$").unwrap()
    });
    let mut events = Vec::new();
    let mut used = 0;
    let mut truncated = false;
    'messages: for message in &data.messages {
        if !matches!(message.role.as_str(), "user" | "assistant") {
            continue;
        }
        let mut seen = HashSet::new();
        for text in crate::transcript::visible_texts(message) {
            let text = crate::query_text::normalize(&text);
            if text.is_empty()
                || IGNORE.is_match(&text)
                || !seen.insert(caseless::default_case_fold_str(&text))
            {
                continue;
            }
            let mut event = Event {
                role: message.role.clone(),
                text,
            };
            let size = event.render().chars().count() + 1;
            if used + size > 12_000 {
                let remaining =
                    12_000_usize.saturating_sub(used + size - event.text.chars().count());
                if remaining > 0 {
                    event.text = event.text.chars().take(remaining).collect();
                    events.push(event);
                }
                truncated = true;
                break 'messages;
            }
            used += size;
            events.push(event);
        }
    }
    let mut chunks = Vec::new();
    let mut chunk = Vec::new();
    let mut size = 0;
    for event in events {
        let event_size = event.render().chars().count() + 1;
        if !chunk.is_empty() && size + event_size > 3200 {
            chunks.push(std::mem::take(&mut chunk));
            size = 0;
        }
        chunk.push(event);
        size += event_size;
    }
    if !chunk.is_empty() {
        chunks.push(chunk);
    }
    (chunks, truncated)
}
